### PR TITLE
feat(cohorts): day and night schedules for cohorts

### PR DESCRIPTION
**Pull request by skoob13**
## Problem

The maximum number of cohorts that can be recalculated daily is 36000 (30 tasks per hour * 50 parallel cohorts * 24 hours). We recently exceeded this number, so cohorts are recalculated once every ~34 hours.

## Changes

Added two separate day and night schedules.

**Day**

Calculate N cohorts every two minutes with a 180-second expiration time. The schedule for the US cluster is from 6 AM to 5:59 PM PT, which is 9 AM to 8:59 PM ET. The schedule for the EU cluster is from 8 AM to 7:59 PM CET.

If N=50, it should calculate 18000 cohorts.

**Night**

Calculate M cohorts every one minute with a 90-second expiration time. The schedule for the US cluster is from 6 PM to 5:59 AM PT, which is 9 PM to 8:59 AM ET. The schedule for the EU cluster is from 8 PM to 7:59 AM CET.

If N=75, it should calculate 54000 cohorts, which is 72000 in total.

Additionally, I added a metric to track cohort staleness in hours.

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Manual testing + running unit tests.

## Additional context

[Slack](https://posthog.slack.com/archives/C0368RPHLQH/p1726767590920189)

### CODE REVIEW

**Comment by anirudhpillai** [comment]
This is great. Though a dumb question, why do we have a smaller expiration time for the night schedule? just because we have more in parallel?

Also how do we limit the total number of concurrent queries to clickhouse. ANALYTICS_QUERIES queue has a limit of 90 concurrent queries to clickhouse, + we'll have 75 parallel now on top (don't think sync_execute which cohort calculation uses, limits parallel queries to Clickhouse). Bumping to 75 should be safe (now that we increased our total CH limit to 200 queries) but just to be sure.

**Comment by skoob13** [comment]
@anirudhpillai

> This is great. Though a dumb question, why do we have a smaller expiration time for the night schedule? just because we have more in parallel?

Yeah, I followed that logic. Generally, the `calculate_cohorts` function should be fast, but we want to prevent tasks from staking if Postgres is unavailable. Only two tasks could be stuck at the same time, so we don't overload Clickhouse if they suddenly "unstuck" at the same time. For example, a maximum of 100 (day) and 150 (night) queries from could appear from cohorts.

> Also how do we limit the total number of concurrent queries to clickhouse caused by cohorts. ANALYTICS_QUERIES queue has a limit of 90 concurrent queries to clickhouse, + we'll have 75 parallel now on top (don't think `sync_execute` which cohort calculation uses, limits parallel queries to Clickhouse). Bumping to 75 should be safe (now that we increased our total CH limit to 200 queries) but just to be sure.

The only limit is the variables I believe. This is something we need to refactor soon because the `ANALYTICS_QUERIES` queue is not in sync with cohorts.

**Review by anirudhpillai — APPROVED** [review]
Makes sense. We'll need to sync up all the queries to prevent exceeding the CH threshold but can do that later.

File: posthog/tasks/scheduled.py [hunk 1]
```diff
@@ -200,11 +200,20 @@ def setup_periodic_tasks(sender: Celery, **kwargs: Any) -> None:
...
     )
 
-    add_periodic_task_with_expiry(
-        sender,
-        120,
+    sender.add_periodic_task(
+        crontab(settings.CALCULATE_COHORTS_DAY_SCHEDULE),
+        calculate_cohort.s(),
+        name="recalculate cohorts",
+        expires=120 * 1.5,
```
Comment (aspicer) [inline posthog/tasks/scheduled.py:207]: Are these expires backwards? (Should day be 60 * 1.5 and night be 120 * 1.5?)

**Comment by aspicer** [comment]
Looks great.

One nit (which might be feature creep): it would be nice to add a feature that causes day schedules to drift to night schedules.

Right now, the only thing that will substantially cause the time a cohort gets generated in the day to drift is if we're not hitting our SLA anymore.

Some sort of feedback that pushes cohorts to run at night as long as night has load for them (making day only for overflow) would be nice.

Not necessarily the best idea but a dumb and simple idea for this would be:
Any cohort that is recomputed during the day has its next recompute time set to 25 hours away instead of 24 horus.

This causes every day calculating cohort to slowly snake around and end up a night calculating cohort after a week or two.

It would probably also be good to apply a little bit of nighttime back pressure to prevent drift into morning and day computation (Night cohort recalculation is set to `24 hours - a little bit that depends on how close to the morning it is`), but this is probably unnecessary, since only a few night cohorts will migrate into being day cohorts every day.

**Pull request merged**